### ISSUE TITLE
Lint `permissions.set_readonly(false)`

### ISSUE THREAD

**Issue by Alexendoo** [rust-lang/rust-clippy#9702]
### What it does

The lint would catch any usage of [`Permissions::set_readonly`](https://doc.rust-lang.org/nightly/std/fs/struct.Permissions.html#method.set_readonly) called with `false`

`.set_readonly(false)` on unix platforms sets the file to be world writable - https://github.com/rust-lang/rust/pull/101644

### Lint Name

permissions_set_readonly_false

### Category

suspicious

### Advantage

The behaviour is surprising, so a lint to inform people that it's probably not doing what they expect would be useful

### Drawbacks

If someone knows about that behaviour and is using it intentionally it would be a false positive, however in that case the `#[allow]` would be a good indicator to others that it is intentional

It's possible that the behaviour will be changed or deprecated upstream before the lint is written

### Example

```rust
let f = File::create("foo.txt")?;
let metadata = f.metadata()?;
let mut permissions = metadata.permissions();

permissions.set_readonly(false);
```

**Comment by dboso** [comment]
Hello

This is my first issue working on clippy, so please bear with me
After having read the documentation and browsed the source code I have made some progress on the issue, unfortunately im stuck on the last part

Currently I have a lint that checks the following conditions
- is a method call
- the callee is of type `std::fs::Permissions`
- the called method is named `set_readonly`
- the method takes one argument of the type bool

The last check I need to perform is to check whether the argumen's value is `false`. This seems like it should be pretty easy to figure out, but I haven't been able to find the way to extract this information

Any hints or pointers to documentation that can help me to understand how to do this is very much appreciated

**Comment by Alexendoo** [comment]
Hi, welcome to Clippy!

The most direct way would be to take a look at the expression's kind to see if it's an [`ExprKind::Lit`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.ExprKind.html#variant.Lit), there you can check that the literal is a [`LitKind::Bool(false)`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/enum.LitKind.html#variant.Bool)

**Comment by dboso** [comment]
Hi, thanks for the reply - very helpful, im learning a lot :-)

For this specific issue (please correct me if im wrong) the input argument to `set_readonly` can be many other things than a literal e.g. `Call`, `MethodCall`, `Unary`, etc..)

Is the idiomatic way to check for each scenario, and identify whether or not they result in a constant expression that always yields `false`

I see there already exists methods to extract type information about expressions like
`cx.typeck_results().expr_ty(receiver)`

Do we have similar methods to extract the value from constant expressions?

**Comment by Alexendoo** [comment]
It could be, there is https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/consts/fn.constant_simple.html

It's not a full const evaluator so it wouldn't be catching `[Method]Calls`, the main thing it'd get here would be `const`s but that isn't super likely to occur

**Comment by Alexendoo** [comment]
Added by #10063
